docs/design.md

```diff
@@ -38,13 +38,112 @@ spec:
 ```
 
 ### Command
+
 Optionally you can override the default command of a container by providing
 `command` in the Substratus resource:
 ```yaml
 spec:
   command: ["train.sh"]
 ```
 
+### Storage
+
+#### Bucket Paths
+
+In Kubernetes the combination of the following fields make up a unique reference to a given object (i.e. database key): `split(.apiVersion, "/")[0] + .kind + .metadata.namespace + .metadata.name`. By storing related artifacts using a similar scheme, restore operations are made trivial: the administrator will not need to worry about backing up the `.status.url` field (as opposed to a scheme where storage location is `.metadata.uid` based - which would be unique for a given in-time instance of an object). When deleting and recreating clusters, if the storage bucket persisted, objects (i.e. Models, Datasets) can simply be re-applied into the new cluster (i.e. [velero](https://velero.io/)/similar is not needed). This plays nicely with GitOps.
+
+Storage scheme:
+
+```
+# Models
+gs://{project_id}-substratus-storage/namespaces/{namespace}/models/{name}/model
+gs://{project_id}-substratus-storage/namespaces/{namespace}/models/{name}/logs
+
+# Datasets
+gs://{project_id}-substratus-storage/namespaces/{namespace}/datasets/{name}/data
+gs://{project_id}-substratus-storage/namespaces/{namespace}/datasets/{name}/logs
+```
+
+NOTE: With this approach, a single bucket could be used to store all Models and Datasets for a given cluster.
+
+Pseudo-reconciler logic for a Model:
+
+```
+if .status.url != "" {
+  return
+}
+
+url := "{bucket}/namespaces/{namespace}/models/{name}"
+if sci.bucketObjectExists(url) {
+  # Use the bucket as the source of truth if .status.url did not exist.
+  updateStatus(url)
+  return
+}
+
+# ... Looks like the model should be imported/trained/etc.
+
+runJob()
+```
+
+#### Mount Points
+
+All mount points will are made under a standardized `/ml` directory which should correspond to the `WORKDIR` of a Dockerfile. This works well for Jupyter notebooks which can be run with `/ml` set as the serving directory: all relevant mounts will be populated on the file explorer sidebar.
+
+The `logs/` directories below are used to store the notebook cell output, python logs, tensorboard stats, etc. These directories are mounted as read-only in Notebooks to explore the output of other background jobs.
+
+##### Dataset (importing)
+
+```
+/ml/params.json  # Populated from .spec.params (also available as env vars).
+
+/ml/data/        # Mounted RW: initially empty dir to write new files
```

Comment:
Should we use `dataset/` instead of `data/` to match `lower(.kind)`?

Reply:
I for one like that adjustment. Better consistency